### DISCUSSION TITLE
About customized package problem

### DISCUSSION THREAD

**Discussion by jczh98** [xmake-io/xmake#3716] [Q&A] [answered]
I have a customized package repo, and i want to add a custom package `A`, the structure of file likes:
```
repo
- packages
-- A
--- xmake.lua
--- port
---- A
---- xmake.lua 
```
the xmake file in A is like:
```
package("A")
    on_load(function (package)
        package:add("defines", "TEST_MACRO")
    end)

    on_install(function (package)
        os.mkdir(package:installdir())
        os.cp(path.join(package:scriptdir(), "port", "A"), ".")
        os.cp(path.join(package:scriptdir(), "port", "xmake.lua"), "xmake.lua")

        local configs = {}        
        import("package.tools.xmake").install(package, configs)
    end)
```

the xmake file in port is like:
```
target("A")
    set_kind("static")
    add_files("A.cpp")
```

My question is that when I use `package:add("defines", "TEST_MACRO")` in xmake.lua in A, the macro is not work.
It only works that when i use `add_defines("TEST_MACRO")` in xmake.lua in port. 
That's why?

**Comment by waruqi** [comment, marked as answer]
> `add_defines` in package

It will only be exported to the user project, not on_install.

if you want to add macros to port/xmake.lua. you need add `add_defines` in port/xmake.lua, not package configuration.

**Reply by jczh98** [reply]
Thanks! It means that `package` is for user project
